### ISSUE TITLE
Backtest and recalibrate HeuristicStrategy prediction weights

### ISSUE THREAD

**Issue by buka4rill** [buka4rill/fpl-bot#3]
## Problem

`HeuristicStrategy`'s scoring weights (form, xG/xA underlying-stats bonus, defensive-contribution bonus, fixture multiplier) are hand-tuned guesses, not backtested — this is called out in the strategy's own header comment, and is no longer theoretical.

A live example on 2026-09-08 demonstrated it concretely: a real chip proposal reported `expectedGain: 175.47` (Bench Boost) — implausible for one gameweek, when even an elite live captain haul tops out around 20-24 points. Hand-recomputing the exact formula against live `bootstrap-static` data confirmed this wasn't an arithmetic bug — the code does exactly what it's defined to do. Three compounding issues in the model itself:

1. FPL's own `form` field can already sit at 8-9+ for an early-season player on a small-sample hot streak, before any bonus is added.
2. The xG/xA underlying-stats bonus and defensive-contribution bonus both add on top of `form`, but elevated `form` is often already *caused by* the same good underlying/defensive numbers — double-counting the same signal rather than correcting for luck (the xG bonus's original intent).
3. The fixture multiplier (up to 1.67×) is multiplicative on top of that already-inflated base, amplifying any overestimate from (1)/(2) rather than gently adjusting it.

This directly undermines every guardrail built on top of the model — `chipRiskPremium` (8 pts) and `OPTIMIZER_HIT_RISK_PREMIUM` are trivial to clear when the underlying numbers are inflated by this much.

## Proposed solution

Deliberately **not** patched with another guessed constant (e.g. capping predicted points at some ceiling) — that repeats the same mistake with a different number.

Real fix: use `ResultsService`'s accumulating predicted-vs-actual history (see #3-adjacent tracking) to:
1. Quantify the actual error/bias in current weights across a real sample of gameweeks.
2. Recalibrate `form`/xG-xA-bonus/defensive-contribution-bonus/fixture-multiplier weights against that data — likely reducing the multiplicative fixture effect and/or de-duplicating the form vs. underlying-stats overlap.
3. Re-verify the risk-premium guardrails (`OPTIMIZER_CHIP_RISK_PREMIUM`, `OPTIMIZER_HIT_RISK_PREMIUM`) still make sense once the base predictions are better calibrated — they may need to shrink once overestimation is fixed, or the fix may make them unnecessary in their current blunt form.

## Timeline

Gated on data, not a calendar date: `ResultsService` needs roughly half a season (15-19 gameweeks) of clean predicted-vs-actual history. Real accumulation only started once prod began running continuously on Fly (2026-09-08); GW4 (deadline 2026-09-12) and earlier are contaminated by manual execution testing on the same account, so the clean count effectively starts at GW5 (deadline 2026-09-18). Projected against the real season calendar, that puts the earliest reasonable start around **GW20-24, roughly January-February 2027**. Re-check periodically via:

```sql
SELECT count(DISTINCT (season, "gameweekId")) FROM proposals WHERE "resultReportedAt" IS NOT NULL;
```

**Comment by buka4rill** [comment]
## Note: how the retrained model should actually be built (not decided, just recorded)

Raised 2026-09-09 while `trained-model.strategy.ts` was open: once this issue's data checkpoint lands and `TrainedModelStrategy` is actually implemented, it's a regression problem (predict points from tabular features — form, xG90/xA90, defensive contribution, fixture difficulty, minutes, position), not an LLM task. No language model involved.

Three options, weighed against this project's existing bias toward keeping everything in-process (Postgres+TypeORM, `javascript-lp-solver` for the ILP rather than a separate solver service):

1. **Stay in TS** — fit a linear/logistic regression or gradient-boosted trees in-process (`ml-regression` or similar). Keeps the no-separate-service pattern, but the JS ML ecosystem is thin beyond linear regression.
2. **Train offline in Python** (scikit-learn/XGBoost/LightGBM) against the accumulated `PlayerSnapshot`/results history, export the fitted model (ONNX, or just serialized weights if it ends up simple enough), and have `TrainedModelStrategy` do inference in TS against that artifact — no Python at runtime, only at training time.
3. **A standalone Python training/serving service** NestJS calls at inference time — most capable tooling, but reintroduces the "separate service" architecture this project has deliberately avoided everywhere else.

Leaning toward **option 2** given the project's track record, but this isn't decided — worth revisiting for real once the GW20-24 data checkpoint actually lands, informed by how complex the winning model ends up being (a plain linear refit might make option 1 fine after all).

**Comment by buka4rill** [comment]
## Research: how other FPL prediction tools calculate expected points

Looked into how established FPL prediction tools/communities compute per-player expected points, to have concrete techniques ready for when this issue's backtesting checkpoint lands (rather than guessing at another hand-tuned constant, same trap the `chipRiskPremium` guardrail and the multiplicative fixture multiplier both fell into).

### Commercial tools
Mostly opaque — Fantasy Football Hub's "Algorithm" and FPL Review don't publish real methodology. The one commercial tool with a public writeup, **FPL Pulse**, uses a gradient-boosted ensemble on rolling stats (points/minutes/xG/xA/ICT) plus bookmaker odds and playing-chance flags, and treats FPL's own official xP as just one input feature rather than the answer itself (reports 0.81 MAE vs 1.05 for raw FPL xP).

### Academic / open-source (more useful — real methodology visible)
- The standard non-proprietary approach **Poisson-simulates matches** from team attack/defense rates, then `xPts = 3·P(win) + 1·P(draw) + goal/assist contribution` — fixture strength combines **additively in log-space**, not as a raw multiplier.
- Several projects (`fpl-predict`, `fpl-ai`) split into **two stages**: will-they-play (minutes/participation) and points-if-played, rather than one flat per-player score.
- Multiple sources explicitly flag `xG`/`xA`/`form` as **collinear**, not independent signals — one dropped xG entirely, others use ridge regression or let a tree model pick the most informative one rather than summing all of them.
- The most rigorous validation protocol found (`smartplayfpl-dastan`) uses **chronological, never-leak-future splits** and per-gameweek rank correlation, not pooled accuracy.

### How this maps onto the over-optimistic-predictions bug this issue exists to fix
Every one of the three compounding causes already diagnosed (see CLAUDE.md's "Live-testing the above surfaced a real prediction-calibration problem" section) has a known fix in this literature:
1. Naive `form + underlyingStatsBonus + defensiveContributionBonus` summation → should be regularized/shrunk, not summed raw — this is exactly the double-counting problem found live.
2. `form` on a small early-season sample → needs regression-to-the-mean/shrinkage before use.
3. Multiplicative fixture multiplier (up to 1.67×) compounding an already-inflated base → should be additive in log-space instead, the way real Poisson projectors do it.

Plus one idea not currently in `HeuristicStrategy` at all: separating "will they play" (minutes/participation) from "how good are they if they play."

None of this changes the plan already decided — wait for real predicted-vs-actual history via `ResultsService` before touching weights — it just gives concrete techniques (shrinkage, log-space fixture adjustment, two-stage modeling, chronological per-gameweek validation) to reach for once that data exists, instead of guessing at new constants again.

**Sources**: [FFH Algorithm page](https://www.fantasyfootballhub.co.uk/the-algorithm-fpl-sky-prediction-planning-toolkit), [FPL Pulse model writeup](https://www.fplpulse.com/blog/fpl-predicted-points-model), [OpenFPL (arXiv)](https://arxiv.org/pdf/2508.09992), [Ramezani & Dinh (arXiv)](https://arxiv.org/pdf/2505.02170), [francisbilham11/fpl-predict](https://github.com/francisbilham11/fpl-predict), [saheedniyi02/fpl-ai](https://github.com/saheedniyi02/fpl-ai), [ritviyer/FPL-Team-Prediction](https://github.com/ritviyer/FPL-Team-Prediction), [qazybekb/smartplayfpl-dastan METHODOLOGY.md](https://github.com/qazybekb/smartplayfpl-dastan/blob/main/docs/METHODOLOGY.md), [R-bloggers FPL model in R](https://www.r-bloggers.com/2021/03/building-a-model-in-r-to-predict-fpl-points/).

---
🤖 Research compiled with [Claude Code](https://claude.com/claude-code)
